include/onnxruntime/core/framework/data_types_internal.h

```diff
@@ -341,25 +345,51 @@ class MLTypeCallDispatcherRet {
   }
 };
 
-// Version of the MLTypeDispatcher that has an input type which is passed through ('carried')
-// as the first type parameter in the call to Fn when dispatching.
-template <typename TCarried, template <typename, typename> class Fn, typename... Types>
-class MLTypeCallDispatcherWithCarriedType {
+// TODO consolidate this with the other MLTypeCallDispatcher classes
+// can add additional methods to cover their usages, but need to update call sites
+template <typename... Types>
+class MLTypeCallDispatcher2 {
+  static_assert(boost::mp11::mp_is_set<TypeList<Types...>>::value,
+                "MLTypeCallDispatcher requires a set of unique types.");
+
   int32_t dt_type_;
 
  public:
-  explicit MLTypeCallDispatcherWithCarriedType(int32_t dt_type) noexcept : dt_type_(dt_type) {}
+  explicit MLTypeCallDispatcher2(int32_t dt_type) noexcept : dt_type_(dt_type) {}
 
-  template <typename... Args>
+  template <template <typename> class Fn, typename... Args>
   void Invoke(Args&&... args) const {
     mltype_dispatcher_internal::CallableDispatchableHelper helper(dt_type_);
-    int results[] = {0, helper.template Invoke<Types>(Fn<TCarried, Types>(), std::forward<Args>(args)...)...};
-    ORT_UNUSED_PARAMETER(results);
-    ORT_ENFORCE(helper.called_ < 2, "Check for duplicate types in MLTypeCallDispatcher");
+
+    std::array<int, sizeof...(Types)>{
+        helper.template Invoke<Types>(Fn<Types>(), std::forward<Args>(args)...)...};
+
+    // avoid "unused parameter" warning for the case where Types is empty
+    std::array<int, sizeof...(Args)>{(ORT_UNUSED_PARAMETER(args), 0)...};
+
```

Comment:
Does an empty Types list occur in valid usage? 

Depending on that answer, should there be a static_assert that it's not empty (if that's not valid), or and 'if' using mp11::mp_empty to handle that case more explicitly?

Reply:
i ran into it with Cast with a single type. there, we have two dispatchers: one with type list SrcTypes and the other with type list (DstTypes - given SrcType). when SrcTypes == DstTypes and they have a single element, the second dispatcher will have an empty type list. i think it's fine for this to be valid. calling Invoke() will fail though.
not sure whether `if(mp_empty<TypeList<Types...>>::value)` would make much difference here other than being notational - the problem is a compiler warning about the args pack. i think both branches would still be "seen" by the compiler. with C++17's if constexpr that could be more explicit though.

---
In reply to: [564866954](https://github.com/microsoft/onnxruntime/pull/6466#discussion_r564866954) [](ancestors = 564866954)

Reply:
ok. do we need to create an array though vs. just doing static_cast<void>(args) (which is what ORT_UNUSED_PARAMETER does)? 

---
In reply to: [564916970](https://github.com/microsoft/onnxruntime/pull/6466#discussion_r564916970) [](ancestors = 564916970,564866954)

Reply:
actually it was a warning for each arg within args. this array is one way to put ORT_UNUSED_PARAMETER on each arg.

---
In reply to: [564921484](https://github.com/microsoft/onnxruntime/pull/6466#discussion_r564921484) [](ancestors = 564921484,564916970,564866954)